{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       ".container {\n",
       "    width: 100%;\n",
       "}\n",
       "div#notebook {\n",
       "    padding-top: 1em;\n",
       "}\n",
       "#header-container {\n",
       "    display: none;\n",
       "}\n",
       "#header-bar {\n",
       "    display: none;\n",
       "}\n",
       "#maintoolbar {\n",
       "    display: none;\n",
       "}\n",
       "#menubar-container {\n",
       "    position: fixed;\n",
       "    margin-top: 0;\n",
       "}\n",
       "#site {\n",
       "    height: auto !important;\n",
       "}\n",
       "</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "%run setup.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|seqid</th>\n",
       "<th>1|type</th>\n",
       "<th>2|start</th>\n",
       "<th>3|stop</th>\n",
       "<th>4|length</th>\n",
       "<th>5|strand</th>\n",
       "<th>6|phase</th>\n",
       "<th>7|parent</th>\n",
       "<th>8|ID</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>chromosome</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td style='text-align: right'>49364325</td>\n",
       "<td style='text-align: right'>49364325</td>\n",
       "<td>.</td>\n",
       "<td>.</td>\n",
       "<td>None</td>\n",
       "<td>2L</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>three_prime_UTR</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>148</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>three_prime_UTR</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>148</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>276</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>276</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>mRNA</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>181305</td>\n",
       "<td style='text-align: right'>23958</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>gene</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>186936</td>\n",
       "<td style='text-align: right'>29589</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>None</td>\n",
       "<td>AGAP004677</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>mRNA</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>186936</td>\n",
       "<td style='text-align: right'>29589</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>128</td>\n",
       "<td>-</td>\n",
       "<td>2</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>AGAP004677-PA</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>128</td>\n",
       "<td>-</td>\n",
       "<td>2</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>AGAP004677-PB</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>157678</td>\n",
       "<td style='text-align: right'>158297</td>\n",
       "<td style='text-align: right'>619</td>\n",
       "<td>-</td>\n",
       "<td>0</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>AGAP004677-PA</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>157678</td>\n",
       "<td style='text-align: right'>158297</td>\n",
       "<td style='text-align: right'>619</td>\n",
       "<td>-</td>\n",
       "<td>0</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>AGAP004677-PB</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>157678</td>\n",
       "<td style='text-align: right'>158297</td>\n",
       "<td style='text-align: right'>619</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>157678</td>\n",
       "<td style='text-align: right'>158297</td>\n",
       "<td style='text-align: right'>619</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>159191</td>\n",
       "<td style='text-align: right'>159356</td>\n",
       "<td style='text-align: right'>165</td>\n",
       "<td>-</td>\n",
       "<td>0</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>AGAP004677-PB</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>159191</td>\n",
       "<td style='text-align: right'>159366</td>\n",
       "<td style='text-align: right'>175</td>\n",
       "<td>-</td>\n",
       "<td>1</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>AGAP004677-PA</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>159191</td>\n",
       "<td style='text-align: right'>159366</td>\n",
       "<td style='text-align: right'>175</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>159191</td>\n",
       "<td style='text-align: right'>159366</td>\n",
       "<td style='text-align: right'>175</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>five_prime_UTR</td>\n",
       "<td style='text-align: right'>159356</td>\n",
       "<td style='text-align: right'>159366</td>\n",
       "<td style='text-align: right'>10</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>180908</td>\n",
       "<td style='text-align: right'>181213</td>\n",
       "<td style='text-align: right'>305</td>\n",
       "<td>-</td>\n",
       "<td>0</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>AGAP004677-PA</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tbl_features = (\n",
    "    etl\n",
    "    .fromgff3(phase2_ar1.geneset_agamp44_fn)\n",
    "    .convert('start', lambda v: v-1)\n",
    "    .rename('end', 'stop')\n",
    "    .unpackdict('attributes', ['Parent', 'ID'])\n",
    "    .rename('Parent', 'parent')\n",
    "    .addfield('length', lambda rec: rec.stop-rec.start, index=5)\n",
    "    .cutout('source', 'score')\n",
    "    .cache()\n",
    ")\n",
    "\n",
    "tbl_features.display(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|seqid</th>\n",
       "<th>1|type</th>\n",
       "<th>2|start</th>\n",
       "<th>3|stop</th>\n",
       "<th>4|length</th>\n",
       "<th>5|strand</th>\n",
       "<th>6|phase</th>\n",
       "<th>7|parent</th>\n",
       "<th>8|ID</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>157678</td>\n",
       "<td style='text-align: right'>55</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>157678</td>\n",
       "<td style='text-align: right'>55</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>158297</td>\n",
       "<td style='text-align: right'>159191</td>\n",
       "<td style='text-align: right'>894</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>158297</td>\n",
       "<td style='text-align: right'>159191</td>\n",
       "<td style='text-align: right'>894</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>159366</td>\n",
       "<td style='text-align: right'>180908</td>\n",
       "<td style='text-align: right'>21542</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>159366</td>\n",
       "<td style='text-align: right'>186859</td>\n",
       "<td style='text-align: right'>27493</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>203924</td>\n",
       "<td style='text-align: right'>203980</td>\n",
       "<td style='text-align: right'>56</td>\n",
       "<td>+</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004678-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>207953</td>\n",
       "<td style='text-align: right'>208392</td>\n",
       "<td style='text-align: right'>439</td>\n",
       "<td>+</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004679-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>208581</td>\n",
       "<td style='text-align: right'>208638</td>\n",
       "<td style='text-align: right'>57</td>\n",
       "<td>+</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004679-RA</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>intron</td>\n",
       "<td style='text-align: right'>208581</td>\n",
       "<td style='text-align: right'>208638</td>\n",
       "<td style='text-align: right'>57</td>\n",
       "<td>+</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004679-RB</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def exons2introns(parent, exons):\n",
    "    exons = list(exons)\n",
    "    seqid = exons[0].seqid\n",
    "    strand = exons[0].strand\n",
    "    type = 'intron'\n",
    "    for i in range(1, len(exons)):\n",
    "        start = exons[i-1].stop\n",
    "        stop = exons[i].start\n",
    "        yield (seqid, type, start, stop, stop-start, strand, '.', parent, None)\n",
    "            \n",
    "\n",
    "tbl_introns = (\n",
    "    tbl_features\n",
    "    .eq('type', 'exon')\n",
    "    .rowgroupmap(key='parent',\n",
    "                 mapper=exons2introns,\n",
    "                 header=['seqid', 'type', 'start', 'stop', 'length', 'strand','phase', 'parent', 'ID'])\n",
    "    .sort(key=('seqid', 'start', 'parent'))\n",
    ")\n",
    "tbl_introns.display(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "lkp_feature_children = tbl_features.recordlookup('parent')\n",
    "lkp_feature_length = tbl_features.lookup('length','type')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gene', 'mRNA', 'mRNA', 'gene', 'mRNA', 'three_prime_UTR']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lkp_feature_length[3000]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (<ipython-input-7-ba4bfb42c590>, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-7-ba4bfb42c590>\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    lkp_feature_children['AGAP004677']no\u001b[0m\n\u001b[0m                                       ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "lkp_feature_children['AGAP004677']no"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcript_length(row):\n",
    "    if row['type'] == 'mRNA':\n",
    "        exons = [f for f in lkp_feature_children[row.ID] if f['type'] == 'exon']\n",
    "        l = sum(e.length for e in exons)\n",
    "        return l\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "\n",
    "lkp_transcript_length = (\n",
    "    tbl_features.eq('type', 'mRNA')\n",
    "    .addfield('transcript_length', transcript_length)\n",
    "    .lookupone('ID', 'transcript_length')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1467"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lkp_transcript_length['AGAP004677-RA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_canonical_transcript(row):\n",
    "    if row['type'] == 'mRNA':\n",
    "        length = lkp_transcript_length[row.ID]\n",
    "        lengths = [lkp_transcript_length[f.ID] for f in lkp_feature_children[row.parent]]\n",
    "        return length == max(lengths)\n",
    "    else:\n",
    "        return None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "lkp_transcript_is_canonical = (\n",
    "    tbl_features\n",
    "    .eq('type', 'mRNA')\n",
    "    .addfield('is_canonical', is_canonical_transcript)\n",
    "    .lookupone('ID', 'is_canonical')\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_canonical(row):\n",
    "    if row['type'] == 'mRNA':\n",
    "        return lkp_transcript_is_canonical[row.ID]\n",
    "    elif row.parent in lkp_transcript_is_canonical:\n",
    "        return lkp_transcript_is_canonical[row.parent]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "    \n",
    "tbl_features_aug = (\n",
    "    tbl_features\n",
    "    .cat(tbl_introns)\n",
    "    .sort(key=('seqid', 'start', 'parent', 'type'), cache=False)\n",
    "    .addfield('n_children', lambda row: len(lkp_feature_children[row.ID]) \n",
    "                                        if row.ID is not None and row.ID in lkp_feature_children\n",
    "                                        else 0)\n",
    "    .addfield('transcript_length', transcript_length)\n",
    "    .addfield('is_canonical', is_canonical)\n",
    "    .cache()\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "lkp_feature_children_aug = tbl_features_aug.recordlookup('parent')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "def first_last(row):\n",
    "    if row['type'] in {'exon', 'intron', 'CDS'}:\n",
    "        first, last = False, False\n",
    "        sibs = [f for f in lkp_feature_children_aug[row.parent]\n",
    "                if f['type'] == row['type']]\n",
    "        if row.strand == '+':\n",
    "            starts = sorted([f.start for f in sibs])\n",
    "            index = starts.index(row.start)\n",
    "        else:\n",
    "            starts = sorted([f.stop for f in sibs], reverse=True)\n",
    "            index = starts.index(row.stop)\n",
    "        if index == 0:\n",
    "            first = True\n",
    "        if index == len(sibs) - 1:\n",
    "            last = True\n",
    "        return first, last\n",
    "    return (None, None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|seqid</th>\n",
       "<th>1|type</th>\n",
       "<th>2|start</th>\n",
       "<th>3|stop</th>\n",
       "<th>4|length</th>\n",
       "<th>5|strand</th>\n",
       "<th>6|phase</th>\n",
       "<th>7|parent</th>\n",
       "<th>8|ID</th>\n",
       "<th>9|n_children</th>\n",
       "<th>10|transcript_length</th>\n",
       "<th>11|is_canonical</th>\n",
       "<th>12|is_first</th>\n",
       "<th>13|is_last</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>chromosome</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td style='text-align: right'>49364325</td>\n",
       "<td style='text-align: right'>49364325</td>\n",
       "<td>.</td>\n",
       "<td>.</td>\n",
       "<td>None</td>\n",
       "<td>2L</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>gene</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>186936</td>\n",
       "<td style='text-align: right'>29589</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>None</td>\n",
       "<td>AGAP004677</td>\n",
       "<td style='text-align: right'>2</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>mRNA</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>181305</td>\n",
       "<td style='text-align: right'>23958</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td style='text-align: right'>10</td>\n",
       "<td style='text-align: right'>1467</td>\n",
       "<td>True</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>mRNA</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>186936</td>\n",
       "<td style='text-align: right'>29589</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td style='text-align: right'>10</td>\n",
       "<td style='text-align: right'>1147</td>\n",
       "<td>False</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>276</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>three_prime_UTR</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>148</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>True</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>276</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>three_prime_UTR</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>148</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>False</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>128</td>\n",
       "<td>-</td>\n",
       "<td>2</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>AGAP004677-PA</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>CDS</td>\n",
       "<td style='text-align: right'>157495</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>128</td>\n",
       "<td>-</td>\n",
       "<td>2</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td>AGAP004677-PB</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>False</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "tbl_features_aug2 = (\n",
    "    tbl_features_aug\n",
    "    .addfield('first_last', first_last)\n",
    "    .unpack('first_last', newfields=['is_first', 'is_last'])\n",
    ")\n",
    "tbl_features_aug2.display(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|check</th>\n",
       "<th>1|count</th>\n",
       "<th>2|frequency</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>True</td>\n",
       "<td style='text-align: right'>12447</td>\n",
       "<td style='text-align: right'>0.8358740178631389</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>False</td>\n",
       "<td style='text-align: right'>2444</td>\n",
       "<td style='text-align: right'>0.1641259821368612</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "+-------+-------+--------------------+\n",
       "| check | count | frequency          |\n",
       "+=======+=======+====================+\n",
       "| True  | 12447 | 0.8358740178631389 |\n",
       "+-------+-------+--------------------+\n",
       "| False |  2444 | 0.1641259821368612 |\n",
       "+-------+-------+--------------------+"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Is 'RA' transcript always canonical?\n",
    "(\n",
    "    tbl_features_aug2\n",
    "    .eq('type', 'mRNA')\n",
    "    .addfield('check', lambda row: row.ID[-2:] == 'RA' and row.is_canonical)\n",
    "    .valuecounts('check')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|n_children</th>\n",
       "<th>1|count</th>\n",
       "<th>2|frequency</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td style='text-align: right'>1</td>\n",
       "<td style='text-align: right'>11775</td>\n",
       "<td style='text-align: right'>0.9040307101727447</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>2</td>\n",
       "<td style='text-align: right'>910</td>\n",
       "<td style='text-align: right'>0.06986564299424185</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>3</td>\n",
       "<td style='text-align: right'>211</td>\n",
       "<td style='text-align: right'>0.016199616122840692</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>4</td>\n",
       "<td style='text-align: right'>74</td>\n",
       "<td style='text-align: right'>0.0056813819577735125</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>5</td>\n",
       "<td style='text-align: right'>26</td>\n",
       "<td style='text-align: right'>0.00199616122840691</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>6</td>\n",
       "<td style='text-align: right'>10</td>\n",
       "<td style='text-align: right'>0.0007677543186180423</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>8</td>\n",
       "<td style='text-align: right'>5</td>\n",
       "<td style='text-align: right'>0.00038387715930902113</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>7</td>\n",
       "<td style='text-align: right'>5</td>\n",
       "<td style='text-align: right'>0.00038387715930902113</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>11</td>\n",
       "<td style='text-align: right'>3</td>\n",
       "<td style='text-align: right'>0.00023032629558541266</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>9</td>\n",
       "<td style='text-align: right'>2</td>\n",
       "<td style='text-align: right'>0.00015355086372360844</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>12</td>\n",
       "<td style='text-align: right'>1</td>\n",
       "<td style='text-align: right'>7.677543186180422e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>10</td>\n",
       "<td style='text-align: right'>1</td>\n",
       "<td style='text-align: right'>7.677543186180422e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>20</td>\n",
       "<td style='text-align: right'>1</td>\n",
       "<td style='text-align: right'>7.677543186180422e-05</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td style='text-align: right'>13</td>\n",
       "<td style='text-align: right'>1</td>\n",
       "<td style='text-align: right'>7.677543186180422e-05</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# How many transcripts do genes have?\n",
    "(\n",
    "    tbl_features_aug2\n",
    "    .select(lambda row: row.type == 'gene')\n",
    "    .valuecounts('n_children')\n",
    "    .displayall()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: created directory '/home/joshjrey/Documents/mosquito_project/outputs/build_blocks'\r\n"
     ]
    }
   ],
   "source": [
    "output_dir = '/home/joshjrey/Documents/mosquito_project/outputs/build_blocks'\n",
    "!mkdir -pv {output_dir}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    tbl_features_aug2\n",
    "    .teetsv(os.path.join(output_dir, 'tbl_features.txt.gz'))\n",
    "    .topickle(os.path.join(output_dir, 'tbl_features.pkl.gz'))\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 4.4M\r\n",
      "-rw-r--r-- 1 joshjrey joshjrey 2.3M May  3 16:12 tbl_features.pkl.gz\r\n",
      "-rw-r--r-- 1 joshjrey joshjrey 2.2M May  3 16:12 tbl_features.txt.gz\r\n"
     ]
    }
   ],
   "source": [
    "!ls -lh {output_dir}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class='petl'>\n",
       "<thead>\n",
       "<tr>\n",
       "<th>0|seqid</th>\n",
       "<th>1|type</th>\n",
       "<th>2|start</th>\n",
       "<th>3|stop</th>\n",
       "<th>4|length</th>\n",
       "<th>5|strand</th>\n",
       "<th>6|phase</th>\n",
       "<th>7|parent</th>\n",
       "<th>8|ID</th>\n",
       "<th>9|n_children</th>\n",
       "<th>10|transcript_length</th>\n",
       "<th>11|is_canonical</th>\n",
       "<th>12|is_first</th>\n",
       "<th>13|is_last</th>\n",
       "</tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>chromosome</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td style='text-align: right'>49364325</td>\n",
       "<td style='text-align: right'>49364325</td>\n",
       "<td>.</td>\n",
       "<td>.</td>\n",
       "<td>None</td>\n",
       "<td>2L</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>gene</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>186936</td>\n",
       "<td style='text-align: right'>29589</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>None</td>\n",
       "<td>AGAP004677</td>\n",
       "<td style='text-align: right'>2</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>mRNA</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>181305</td>\n",
       "<td style='text-align: right'>23958</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td style='text-align: right'>10</td>\n",
       "<td style='text-align: right'>1467</td>\n",
       "<td>True</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>mRNA</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>186936</td>\n",
       "<td style='text-align: right'>29589</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677</td>\n",
       "<td>AGAP004677-RB</td>\n",
       "<td style='text-align: right'>10</td>\n",
       "<td style='text-align: right'>1147</td>\n",
       "<td>False</td>\n",
       "<td>None</td>\n",
       "<td>None</td>\n",
       "</tr>\n",
       "<tr>\n",
       "<td>2L</td>\n",
       "<td>exon</td>\n",
       "<td style='text-align: right'>157347</td>\n",
       "<td style='text-align: right'>157623</td>\n",
       "<td style='text-align: right'>276</td>\n",
       "<td>-</td>\n",
       "<td>.</td>\n",
       "<td>AGAP004677-RA</td>\n",
       "<td>None</td>\n",
       "<td style='text-align: right'>0</td>\n",
       "<td>None</td>\n",
       "<td>True</td>\n",
       "<td>False</td>\n",
       "<td>True</td>\n",
       "</tr>\n",
       "</tbody>\n",
       "</table>\n",
       "<p><strong>...</strong></p>"
      ],
      "text/plain": [
       "+-------+--------------+--------+----------+----------+--------+-------+-----------------+-----------------+------------+-------------------+--------------+----------+---------+\n",
       "| seqid | type         | start  | stop     | length   | strand | phase | parent          | ID              | n_children | transcript_length | is_canonical | is_first | is_last |\n",
       "+=======+==============+========+==========+==========+========+=======+=================+=================+============+===================+==============+==========+=========+\n",
       "| '2L'  | 'chromosome' |      0 | 49364325 | 49364325 | '.'    | '.'   | None            | '2L'            |          0 | None              | None         | None     | None    |\n",
       "+-------+--------------+--------+----------+----------+--------+-------+-----------------+-----------------+------------+-------------------+--------------+----------+---------+\n",
       "| '2L'  | 'gene'       | 157347 |   186936 |    29589 | '-'    | '.'   | None            | 'AGAP004677'    |          2 | None              | None         | None     | None    |\n",
       "+-------+--------------+--------+----------+----------+--------+-------+-----------------+-----------------+------------+-------------------+--------------+----------+---------+\n",
       "| '2L'  | 'mRNA'       | 157347 |   181305 |    23958 | '-'    | '.'   | 'AGAP004677'    | 'AGAP004677-RA' |         10 |              1467 | True         | None     | None    |\n",
       "+-------+--------------+--------+----------+----------+--------+-------+-----------------+-----------------+------------+-------------------+--------------+----------+---------+\n",
       "| '2L'  | 'mRNA'       | 157347 |   186936 |    29589 | '-'    | '.'   | 'AGAP004677'    | 'AGAP004677-RB' |         10 |              1147 | False        | None     | None    |\n",
       "+-------+--------------+--------+----------+----------+--------+-------+-----------------+-----------------+------------+-------------------+--------------+----------+---------+\n",
       "| '2L'  | 'exon'       | 157347 |   157623 |      276 | '-'    | '.'   | 'AGAP004677-RA' | None            |          0 | None              | True         | False    | True    |\n",
       "+-------+--------------+--------+----------+----------+--------+-------+-----------------+-----------------+------------+-------------------+--------------+----------+---------+\n",
       "..."
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "etl.frompickle(os.path.join(output_dir, 'tbl_features.pkl.gz'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Mos-proj",
   "language": "python",
   "name": "mos-proj"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
